java/code/src/com/suse/manager/reactor/messaging/RegisterMinionEventMessageAction.java

```diff
@@ -127,147 +133,175 @@ public void execute(EventMessage msg) {
-                                                  String machineId,
-                                                  Optional<User> creator,
-                                                  boolean isSaltSSH, ValueMap grains) {
-        Optional<MinionServer> optMinion = MinionServerFactory.findByMachineId(machineId);
-        if (optMinion.isPresent()) {
-            MinionServer registeredMinion = optMinion.get();
-            String oldMinionId = registeredMinion.getMinionId();
-
-            if (!minionId.equals(oldMinionId)) {
-                LOG.warn("Minion '" + oldMinionId + "' already registered, updating " +
-                        "profile to '" + minionId + "' [" + machineId + "]");
-                registeredMinion.setName(minionId);
-                registeredMinion.setMinionId(minionId);
-                ServerFactory.save(registeredMinion);
-                SystemManager.addHistoryEvent(registeredMinion, "Duplicate Machine ID", "Minion '" +
-                        oldMinionId + "' has been updated to '" + minionId + "'");
-
-                if (!minionId.equals(oldMinionId)) {
-                    SALT_SERVICE.deleteKey(oldMinionId);
-                }
+    private void applySaltBootStates(String minionId, MinionServer registeredMinion, boolean saltbootInitrd) {
+        // Saltboot treatment
+        // HACK: try to guess if the minion is a retail minion based on its groups.
+        // This way we don't need to call grains for each register minion event.
+        if (isRetailMinion(registeredMinion)) {
+            if (saltbootInitrd) {
+                // if we have the "saltboot_initrd" grain we want to re-deploy an image via saltboot,
+                LOG.info("Applying saltboot for minion " + minionId);
+                applySaltboot(registeredMinion);
             }
+        }
+    }
 
-            // Saltboot treatment
-            // HACK: try to guess if the minion is a retail minion based on its groups.
-            // This way we don't need to call grains for each register minion event.
-            if (isRetailMinion(registeredMinion)) {
-                if (grains.getOptionalAsBoolean("saltboot_initrd").orElse(false)) {
-                    // if we have the "saltboot_initrd" grain we want to re-deploy an image via saltboot,
-                    LOG.info("Applying saltboot for minion " + minionId);
-                    applySaltboot(registeredMinion);
-                }
+    /**
+     * Reactivate the system
+     * @param minionId minion id of the minion
+     * @param machineId machine_id of the minion
+     * @param reActivationKey valid Reaction key
+     */
+    private void reactivateSystem(String minionId, String machineId, String reActivationKey) {
+        of(ActivationKeyFactory.lookupByKey(reActivationKey)).ifPresent(ak -> {
+            if (Objects.nonNull(ak.getKickstartSession())) {
+                ak.getKickstartSession().markComplete("Installation completed.");
```

Comment:
One extra twist: in original code, the Kickstart session is marked complete as soon as the activation key is found in the database and has a session. In your code, this is done only if the activation key is a reactivation key.

Thinking about it a bit, completing a Kickstart session has nothing to do with reactivations, and actually nothing to do with registration either. It has more to do with normal rebooting / minion start event: if we see a minion that was being autoinstalled come up, then the autoinstallation has ended!

I guess this `if` block has to be moved outside this method.

Reply:
Thank you very much @moio. I would have surely introduced some bugs after this change without your review. I have added a new commit, please have a look.